### PR TITLE
Select, scatter, gather, GNN optimization

### PR DESCRIPTION
**Pull request by Marco-Christiani**

### CODE REVIEW

**Comment by Marco-Christiani** [comment]
@laityjet

File: src/ndtensor.zig [hunk 1]
```diff
@@ -1231,33 +1286,201 @@ pub fn NDTensor(comptime T: type) type {
...
+                        else => |o| break :blk @tagName(o),
                     }
                 };
-                std.debug.print("{?s}\n", .{symbol});
+                std.debug.print(": {?s}\n", .{symbol});
             }
-            var next_children = self.child_iterator() orelse return;
-            while (next_children.next()) |elem| elem.print_arrows();
+            var next_children = self.node.child_iterator() orelse return;
+            while (next_children.next()) |elem| elem.upcast(Self).print_arrows();
```
Comment (andrewCodeDev) [inline src/ndtensor.zig:1483]: I think we can move `print_arrows` into the `Node` object because at that level, it doesn't have to care about the type being casted to - it just uses the label and the op tag - we could add the `op` tag to the node as well.

File: examples/gcn/src/model.zig [hunk 2]
```diff
@@ -121,37 +123,185 @@ pub fn GraphConvLayer(comptime T: type) type {
...
                     .label = "conv_adj_mat",
                     .acquired = true,
                 });
-                errdefer adj_mat.deinit();
+                // errdefer adj_mat.deinit();
                 const adj_mat_data = adj_mat.get_data();
 
-                const deg = try self.device.mem_scratch(T, n_node);
-                self.device.mem_fill(T, deg, 1); // the initial is 1 (self loop)
+                const deg = try zg.NDArray(T).scratch(&.{n_node}, self.device);
```
Comment (andrewCodeDev) [inline examples/gcn/src/model.zig:144]: See my comments about `NDArray.scratch` below. 

To make this device safe, we could allocate some memory with a `HostDevice` created within this function and then use the incoming device to transfer the memory over. This function explicitly allocates, so I see no problem with doing it this way since this is all a one-time set up. I recommend using host device over a global allocator (like `std.heap.smp_allocator`) because the error returned will be `TensorOOM`.

**Review by andrewCodeDev — DISMISSED** [review]
Nice work! A lot of new operations.

File: src/ndtensor.zig (hunk 1 above)
Comment (andrewCodeDev) [reply src/ndtensor.zig:1483]: Doesn't have to be this PR.

File: src/ndarray.zig [hunk 3]
```diff
@@ -50,6 +50,13 @@ pub fn NDArray(comptime T: type) type {
             device.mem_free(self.data);
         }
 
+        pub fn scratch(shape: []const usize, device: DeviceReference) !Self {
```
[the remark this answers is not in the document]
Comment (Marco-Christiani) [reply src/ndarray.zig:53]: I realized this after accidentally shooting myself in the foot. I agree, although it seems the caching allocator may be a large work item. Also, we need to do a better job to discourage people accessing memory or knowing about NDArray.data. However, probably the right way to accomplish that is to remove all instances of this from the examples as we add ops. 

What do you think?

File: src/ndtensor.zig [hunk 4]
```diff
@@ -1231,33 +1286,201 @@ pub fn NDTensor(comptime T: type) type {
...
+                pub fn backward(y: *Self, children: *Node.Children, ctx: *@This()) !void {
+                    defer y.device.mem_free(ctx.offsets);
+
+                    const src_tensor = children.get_bwd_upcast(Self, 0) orelse return;
+                    const grad_output = y.assume_grad_data();
+                    const grad_src = try src_tensor.ensure_grad_data(0);
+
+                    // Gather gradients from scattered positions
+                    // take to invert scatter_add
+                    src_tensor.device.mem_take(T, grad_output, ctx.offsets, grad_src);
```
[the remark this answers is not in the document]
Comment (Marco-Christiani) [reply src/ndtensor.zig:1350]: Yes this isn't hooked up to the right kernel yet but this is a good note so I don't forget.

File: src/ndtensor.zig (hunk 1 above)
Comment (Marco-Christiani) [reply src/ndtensor.zig:1483]: The Op tag should go! Haha. I had an idea for that, agreed it should be a different PR for sure... We should move the tracing thing from utils into Node or maintain it. This print arrows thing is the hacky way I get a plot of the graph while debugging.

File: examples/gcn/src/model.zig [hunk 5]
```diff
@@ -38,9 +39,14 @@ pub fn GCN(comptime T: type) type {
                 }
             }
             try zg.nn.relu_(T, c1);
+
+            if (zg.rt_grad_enabled) std.debug.assert(c1.requires_grad());
```
[the remark this answers is not in the document]
Comment (Marco-Christiani) [reply examples/gcn/src/model.zig:43]: Debug code

**Comment by Marco-Christiani** [comment]
- [ ] move loss functions down to device API

File: src/ndtensor.zig [hunk 6]
```diff
@@ -227,7 +233,7 @@ pub fn NDTensor(comptime T: type) type {
         }
 
         pub fn get_label(self: *const Self) ?[]const u8 {
```
[the remark this answers is not in the document]
Comment (Marco-Christiani) [reply src/ndtensor.zig]: Done in `node.zig`

File: src/ndtensor.zig (hunk 4 above)
[the remark this answers is not in the document]
Comment (Marco-Christiani) [reply src/ndtensor.zig]: Done
